src/hotspot/share/services/heapDumper.cpp

```diff
@@ -608,20 +587,287 @@ void DumpWriter::start_sub_record(u1 tag, u4 len) {
   write_u1(tag);
 }
 
-void DumpWriter::end_sub_record() {
+void AbstractDumpWriter::end_sub_record() {
   assert(_in_dump_segment, "must be in dump segment");
   assert(_sub_record_left == 0, "sub-record not written completely");
   assert(!_sub_record_ended, "Must not have ended yet");
   debug_only(_sub_record_ended = true);
 }
 
+// Supports I/O operations for a dump
+
+class DumpWriter : public AbstractDumpWriter {
+ private:
+  CompressionBackend _backend; // Does the actual writing.
+ protected:
+  virtual void flush(bool force = false);
+
+ public:
+  // Takes ownership of the writer and compressor.
+  DumpWriter(AbstractWriter* writer, AbstractCompressor* compressor);
+
+  // total number of bytes written to the disk
+  virtual julong bytes_written() const          { return (julong) _backend.get_written(); }
+
+  virtual char const* error() const             { return _backend.error(); }
+
+  // Called by threads used for parallel writing.
+  void writer_loop()                    { _backend.thread_loop(false); }
+  // Called when finished to release the threads.
+  virtual void deactivate()             { flush(); _backend.deactivate(); }
+  // Get the backend pointer, used by parallel dump writer.
+  CompressionBackend* backend_ptr() { return &_backend; }
+};
+
+// Check for error after constructing the object and destroy it in case of an error.
+DumpWriter::DumpWriter(AbstractWriter* writer, AbstractCompressor* compressor) :
+  AbstractDumpWriter(),
+  _backend(writer, compressor, io_buffer_max_size, io_buffer_max_waste) {
+  flush();
+}
+
+// flush any buffered bytes to the file
+void DumpWriter::flush(bool force) {
+  _backend.get_new_buffer(&_buffer, &_pos, &_size, force);
+}
+
+struct ParWriterBufferQueueElem {
+  char* _buffer;
+  size_t _used;
+  ParWriterBufferQueueElem* _next;
+};
+
+class ParWriterBufferQueue : public CHeapObj<mtInternal> {
+ private:
+  ParWriterBufferQueueElem* _head;
+  ParWriterBufferQueueElem* _tail;
+  uint _length;
+ public:
+  ParWriterBufferQueue() : _head(NULL), _tail(NULL), _length(0) { }
+
+  void enqueue(ParWriterBufferQueueElem* entry) {
+    if (_head == NULL) {
+      assert(is_empty() && _tail == NULL, "Sanity check");
```

Comment:
When I have run the code with assertions enabled, this one failed since _tail was not NULL.